### PR TITLE
[Feature] add mean pool (embed) feature into generate API

### PR DESCRIPTION
**Pull request by zzaebok**
## Purpose

Currently, vLLM doesn't officially support returning both the output text and the pooled vector, which leads to run two vLLM instances to get both results (very inefficient).

However, there are huge community demand for this feature:
- https://github.com/vllm-project/vllm/issues/5449
- https://github.com/vllm-project/vllm/issues/37971
- https://github.com/vllm-project/vllm/issues/11905
- https://github.com/vllm-project/vllm/issues/11577

The issues were closed due to recent feature `extract_hidden_states`. **However**, it is piggyback on speculative decoding (which is extremely slow and not support online serve). In many production cases, getting both output text and embedding vector are very common and useful (e.g., retrieval of user content)

### Implementation details

Implementations are included in model runner. Please follow the commits.

Added `--enable-return-embed` for server argument. For the client request, you can both opt-in / out the returning embedding vector via `extra_args["return_embed"]`. The runner first seeds its running sum from any prefix-cache hits and accumulates each step's prompt-token hidden states into it. When the prompt is fully observed, the mean is finalized, copied to host, and stashed on the request. The scheduler ships it as `kv_transfer_params["embed"]` on the request's final output, alongside the generated text.

Currently, the input args and output format is set temporarily (request is set using `extra_args` and return with `kv_transfer_params`, since it requires changing multiple files). **Will update the schema correctly after maintainers' review and concensus.**

**Current limitation**
- Does not support stream
- Does not support multi turn

However those functionalities is not necessary with the usecase. Normal usecase is getting single-turn output text and embedding vector.

## Test Plan & Test Results

GPU: A100

**1. Added unit tests to compare output with vllm embed**

output of mean pooled vector and output of `embed` endpoint shows
**>= 0.999 cosine similarity**

**2. Checked online serve and response**

```bash
vllm serve meta-llama/Llama-3.2-3B-Instruct \                                                                                          
    --max-model-len 128 --enforce-eager --dtype bfloat16 --port 8765 \
    --enable-return-embed
```

Tested with API calls `/v1/completions` and `/v1/chat/completions` with additional arguments, so that both kinds of requests are allowed (request embed or not)

```bash
curl -X POST http://localhost:8765/v1/completions \                                                                                    
  -H "Content-Type: application/json" \                                                                                                
  -d '{                                                                                                                                
    "model": "meta-llama/Llama-3.2-3B-Instruct",                                                                                       
    "prompt": "The quick brown fox jumps over the lazy dog.",                                                                          
    "max_tokens": 8,                                                                                                                   
    "temperature": 0.0,                                                                                                                
    "vllm_xargs": {"return_embed": 1}                                                                                                  
  }'
```

Output

```bash
{
  "id": "cmpl-...",
  "object": "text_completion",
  "created": 1745382...,
  "model": "meta-llama/Llama-3.2-3B-Instruct",
  "choices": [
    {
      "index": 0,
      "text": " This is a well-known pangram,",
      "logprobs": null,
      "finish_reason": "length",
      "stop_reason": null,
      "prompt_logprobs": null,
      "token_ids": null
    }
  ],
  "usage": {"prompt_tokens": 11, "completion_tokens": 8, "total_tokens": 19},
  "kv_transfer_params": {
    "embed": [-1.1253883838653564, -0.5150923728942871, 2.4300427436828613, -1.70556640625, ...3068 more floats...,
 0.7966974377632141, -0.27256083488464355, -0.4936384856700897, 0.25064921379089355]
  }
}
```

---
<details>
<summary> Essential Elements of an Effective PR Description Checklist </summary>

- [x] The purpose of the PR, such as "Fix some issue (link existing issues this PR will resolve)".
- [x] The test plan, such as providing test command.
- [x] The test results, such as pasting the results comparison before and after, or e2e results
- [ ] (Optional) The necessary documentation update, such as updating `supported_models.md` and `examples` for a new model.
- [ ] (Optional) Release notes update. If your change is user facing, please update the release notes draft in the [Google Doc](https://docs.google.com/document/d/1YyVqrgX4gHTtrstbq8oWUImOyPCKSGnJ7xtTpmXzlRs/edit?tab=t.0).
</details>

### CODE REVIEW

File: vllm/v1/worker/aux_pool_state.py [hunk 1]
```diff
@@ -0,0 +1,178 @@
...
+        Initialize with a request's running sum for
+        prefix-cache and chunked-prefill
+        """
+        assert self._block_aux_sums is not None
+        running_sum = torch.zeros(
+            self._hidden_size, dtype=torch.float32, device=self._device
+        )
+        if num_computed_tokens > 0 and cached_block_ids:
+            # pre-calculated prefix blocks
+            num_full_cached = num_computed_tokens // self._block_size
```
[the remark this answers is not in the document]
Comment (zzaebok) [reply vllm/v1/worker/aux_pool_state.py:98]: It is not a bug on v1. The prefix cache enforces num_computed_tokens % block_size == 0

See the docstring and NOTE of `KVCacheManager.get_computed_blocks` in `vllm/v1/core/kv_cache_manager.py`:

> Note that the computed blocks must be full.

> "allocate_slots() requires num_computed_tokens to be block-size aligned."

**Review by NickLucche — COMMENTED** [review]
cc @noooop

**Comment by noooop** [comment]
I have tried many approaches, but any attempt to add pooling support for generation tasks will lead to unnecessary complexity for runners and entrypoints. 

vLLM is primarily for generation, so let's not support such features in-tree and keep vLLM simple.

https://github.com/vllm-project/vllm/pull/30672

**Comment by zzaebok** [comment]
@noooop Thanks for the review!

I totally understand vLLM's main feature is generation, yet it still supports pooling widely. And running an additional instance for a mean-pooling vector just increased 2x redundant computation, leading to a huge inefficiency.

Since there is a huge demand from users for such feature, I still believe relevant additional support should be made. Currently, related issues are closed as 'solved' by https://github.com/vllm-project/vllm/issues/33118. However **it is not**. Very slow piggybacking on spec decode, not supporting online serve.

To minimize the overhead from runner side and maintain simplicity, I have implemented **kv connector version** (**just like** https://github.com/vllm-project/vllm/issues/33118) before: https://github.com/vllm-project/vllm/pull/38565. But it requires model side change change too (due to last hidden states problem).

Can you please re-consider implementing this feature, either in runner side or in other component side? Please let me know if you have any questions or ideas.

**Comment by noooop** [comment]
cc @DarkLight1337 

Do you think it is necessary to support this feature in-tree?

**Comment by DarkLight1337** [comment]
You and I both know that @WoosukKwon would be against this as it complicates the core logic. I think we should keep this out of tree (or at least outside of GPU runner logic)

**Comment by zzaebok** [comment]
@noooop @DarkLight1337 Then, what do you guys think of this connector approach: https://github.com/vllm-project/vllm/pull/38565? It can merge the feature cleanly without touching the core logic (runner).

Only one problem: last hidden states are not captured correctly, so temporarily modified model code (see llama.py and qwen.py for example). But let me investigate more deeply if we could detour this path if you are okay with this approach.
